Cont. - Add Proto Message Extraction http filter

dchakarwarti opened the pull request:
**Commit Message: Cont. - Add Proto Message Extraction http filter**

**Additional Description:** 

The Proto Message Extraction Filter supports extracting gRPC requests/responses (proto messages) as `google.protobuf.Struct` and storing results in the dynamic metadata `envoy.filters.http.proto_message_scrubbing` for later access.

[Filter Config]( https://github.com/envoyproxy/envoy/blob/acb0626d57c74a6b027ac0a38f0139914c10b614/api/envoy/extensions/filters/http/proto_message_logging/v3/config.proto): Submitted in PR #31735 

Please see the proto comments for this filter's behavior. Thanks Yan (@yanavlasov) for help in reviewing our design inside google.

This PR is in continuation with #34955 in which we added logging util lib.

**Risk Level:** Low

**Testing:** Added unit test and integration test

Docs Changes: Nil
Release Notes: Nil
Platform Specific Features: Nil

mattklein123 commented:
What is the background on this? This is a 6K line PR with no context. Thank you.

/wait-any

dchakarwarti commented:
> What is the background on this? This is a 6K line PR with no context. Thank you.
> 
> /wait-any

Apologies. I have added more context in the PR description.

The design for the PML filter is approved inside Google. Yan (@yanavlasov) is out sponsor for the filter.

Thank you!

dchakarwarti commented:
/assign @yanavlasov

shuoyang2016 commented on api/envoy/extensions/filters/http/proto_message_logging/v3/config.proto:
```diff
@@ -15,20 +15,26 @@ option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/fil
+//
+// ProtoMessageLogging filter supports logging scrubbed gRPC
+// requests/responses(proto messages) as google.protobuf.Struct and storing
+// results in the dynamic metadata `envoy.filters.http.proto_message_logging`
```
Do we want to call it `proto_message_scrubbing`? I think what it does it to scrub and store in the dynamic metadata. The scrubbed result can be used not only for logging?

dchakarwarti replied:
Makes sense. I modified the filter name to `Proto Message Scrubbing`.

Additionally, modified the `logging` / `auditing` to `scrubbing` in filter config and other places.

Thank you!

shuoyang2016 commented on source/extensions/filters/http/proto_message_logging/extractor_impl.cc:
```diff
@@ -0,0 +1,123 @@
+ABSL_CONST_INIT const char* const kTypeServiceBaseUrl = "type.googleapis.com";
+
+void Extract(ProtoScrubberInterface& scrubber, Protobuf::field_extraction::MessageData& message,
+             std::vector<AuditMetadata>& vect) {
```
Can it be called ScrubbedMessageMetadata etc? I think later it can be used for other API transformation and API filtering?

dchakarwarti replied:
Yes, right. Done.

mattklein123 commented on docs/root/configuration/http/http_filters/proto_message_logging_filter.rst:
```diff
@@ -0,0 +1,8 @@
+.. _config_http_filters_proto_message_logging:
+
+Proto Message Logging
+=====================
+
+* gRPC :ref:`architecture overview <arch_overview_grpc>`
+* This filter should be configured with the type URL ``type.googleapis.com/envoy.extensions.filters.http.proto_message_logging.v3.ProtoMessageLoggingConfig``.
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.proto_message_logging.v3.ProtoMessageLoggingConfig>`.
```
Can you add some actual documentation here on what this filter does, why someone would want to use it, and maybe a short config example? Thank you.

/wait

dchakarwarti replied:
Actually, the [config proto](https://github.com/envoyproxy/envoy/blob/06a20dd6ec9f5bd7c6a9e75cc1f54ee0396e7699/api/envoy/extensions/filters/http/proto_message_logging/v3/config.proto) added in #31735 for the filter contains additional details and an example config.

I have verified the doc generation for the v3 API reference using the following command:

```
$ bazel run --//tools/tarball:target=//docs:html //tools/tarball:unpack "$PWD"/generated/docs/
```

IIUC, this will generate the API spec URL under `https://www.envoyproxy.io/docs/envoy/latest/_sources/api-v3/extensions/filters/http/`. This is similar to what we have for the [GFX filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/grpc_field_extraction/v3/config.proto).

Please let me know if we want additional details here.

Thank you!

mattklein123 replied:
Yes please add a least some high level description of what this filter does and why someone would want to use it and then ref links to other places as needed, thanks.

/wait

dchakarwarti replied:
Sure, I have added description, use case and references. Please let me know if it looks good.

Thanks!

mattklein123 commented on docs/root/configuration/http/http_filters/proto_message_scrubbing_filter.rst:
```diff
@@ -0,0 +1,106 @@
+* This filter should be configured with the type URL ``type.googleapis.com/envoy.extensions.filters.http.proto_message_scrubbing.v3.ProtoMessageScrubbingConfig``.
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.proto_message_scrubbing.v3.ProtoMessageScrubbingConfig>`.
+
+ProtoMessageScrubbing filter supports scrubbing scrubbed gRPC
+requests/responses (proto messages) as google.protobuf.Struct and storing
+results in the dynamic metadata `envoy.filters.http.proto_message_scrubbing`
+for later access.
```
```suggestion
ProtoMessageScrubbing filter supports scrubbing gRPC
requests/responses (proto messages) into google.protobuf.Struct and storing
results in the dynamic metadata `envoy.filters.http.proto_message_scrubbing`
for later access.
```

Make this same change in the API docs also.

dchakarwarti replied:
Sure, done.

Thank you!

mattklein123 reviewed (dismissed):
High level changes LGTM. Assigning to @yanavlasov for code review and merge.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/filter.cc:
```diff
@@ -0,0 +1,434 @@
+
+    extractor_->processResponse(*stream_message->message());
+
+    if (end_stream) {
```
What if there's trailers?

dchakarwarti replied:
Good point. We do not need to check `end_stream` here. Message Converter handles that logic. Removed the if condition from request/response path.

Also, added a test case with `end_stream=false` to indicate trailers' presence.

Thanks for noticing!

shuoyang2016 reviewed (commented):
Reviewed filter logic.

Will review scrubber logic shortly.

yanavlasov commented:
It looks like tests do not provide sufficient coverage.
Error: https://dev.azure.com/cncf/envoy/_build/results?buildId=178046&view=logs&j=4930ecaf-18f4-5b3c-dea3-309729c3b3ae&t=573d8780-d7b9-52e3-b4e0-a89886b0b9ff&l=5237

Cov report: https://storage.googleapis.com/envoy-pr/35689/coverage/index.html

/wait

dchakarwarti commented on source/extensions/filters/http/proto_message_scrubbing/extractor_impl.cc:
```diff
@@ -0,0 +1,126 @@
+void Extract(ProtoScrubberInterface& scrubber, Protobuf::field_extraction::MessageData& message,
+             std::vector<ScrubbedMessageMetadata>& vect) {
+  ScrubbedMessageMetadata data = scrubber.ScrubMessage(message);
+  LOG(INFO) << "Extracted scrubbed fields: " << data.scrubbed_message;
```
Sure, done!

dchakarwarti commented on source/extensions/filters/http/proto_message_scrubbing/filter.cc:
```diff
@@ -0,0 +1,434 @@
+      grpc_status, rc_detail);
+
+  encoder_callbacks_->streamInfo().setResponseFlag(
+      Envoy::StreamInfo::CoreResponseFlag::UnauthorizedExternalService);
```
This is derived from SCv2 filter. I do not fully understand why we use `UnauthorizedExternalService` for rejecting response in there. 

In our case, the response is rejected due to "not receiving enough data to form a message". However, I don't see any other flag explicitly mentions insufficient data or malformed response directly. Reference: [envoy/stream_info/stream_info.h](https://github.com/envoyproxy/envoy/blob/c61838567f214812b991f0c76db935a54300d94a/envoy/stream_info/stream_info.h#L39-L99)

shuoyang2016 replied:
Not sure. We can consult Yan otherwise driving from SCv2 filter SG.

dchakarwarti replied:
cc: @yanavlasov Do you have any suggestions here?

dchakarwarti commented on test/extensions/filters/http/proto_message_scrubbing/filter_test.cc:
```diff
@@ -0,0 +1,1837 @@
+}
+
+TEST_F(FilterTestExtractOk, EmptyFields) {
+  setUp();
```
Actually, no. We would need to call it explicitly.

Similar is the case in GFX filter test as well. [filter_test.cc](https://github.com/envoyproxy/envoy/blob/c61838567f214812b991f0c76db935a54300d94a/test/extensions/filters/http/grpc_field_extraction/filter_test.cc#L103)

dchakarwarti commented on test/extensions/filters/http/proto_message_scrubbing/filter_test.cc:
```diff
@@ -0,0 +1,1837 @@
+
+constexpr absl::string_view kFilterName = "envoy.filters.http.proto_message_scrubbing";
+
+constexpr absl::string_view kExpectedRequestMetadata = R"pb(
```
Done.

dchakarwarti commented on test/extensions/filters/http/proto_message_scrubbing/filter_test.cc:
```diff
@@ -0,0 +1,1837 @@
+            filter_->decodeHeaders(req_headers, false));
+}
+
+TEST_F(FilterTestExtractRejected, NothingToExtract) {
```
Yes, missed it.

It should be a pass through. Added the test case under `FilterTestPassThrough`.

dchakarwarti commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/proto_scrubber.h:
```diff
@@ -0,0 +1,70 @@
+  // rehash behavior.
+  std::map<ScrubbedMessageDirective, ProtobufWkt::FieldMask> directives_mapping_;
+  std::function<const ::Envoy::ProtobufWkt::Type*(const std::string&)> type_finder_;
+  std::unique_ptr<proto_processing_lib::proto_scrubber::CloudAuditLogFieldChecker> field_checker_;
```
Actually, the `CloudAuditLogFieldChecker` is part of proto processing lib from [grpc-ecosystem](https://github.com/grpc-ecosystem/proto_processing_lib/tree/main/proto_processing_lib/proto_scrubber).

I will make the change in the repo and update the SHA and names here.

dchakarwarti replied:
Done. Updated the SHA and replaced the name from `CloudAuditLogFieldChecker` to `FieldMaskPathChecker`.

dchakarwarti commented on test/extensions/filters/http/proto_message_scrubbing/filter_test.cc:
```diff
@@ -0,0 +1,1837 @@
+  CreateApiKeyRequest request = makeCreateApiKeyRequest();
+  Envoy::Buffer::InstancePtr request_data = Envoy::Grpc::Common::serializeToGrpcFrame(request);
+
+  EXPECT_DEATH(filter_->decodeData(*request_data, true), ".*");
```
Thanks for noticing.

The `SCRUB_REDACT` field should be used with `struct_value` fields. In this case, we are trying to use the tag with string field.

I have included this check while parsing the config and extractor initialization now. Added this test case in `filter_config_test.cc`.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/scrubbing_util.h:
```diff
@@ -124,7 +126,18 @@ IsMessageFieldPathPresent(const Protobuf::Type& type,
                           const Protobuf::field_extraction::MessageData& message);
 
-} // namespace ProtoMessageLogging
+absl::Status ConvertToStruct(const Protobuf::field_extraction::MessageData& message,
```
nit: Add a comment for this method?

dchakarwarti replied:
Sure, done.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/scrubbing_util.h:
```diff
@@ -124,7 +126,18 @@ IsMessageFieldPathPresent(const Protobuf::Type& type,
+                             ::Envoy::Protobuf::util::TypeResolver* type_resolver,
+                             ::Envoy::ProtobufWkt::Struct* message_struct);
+
+bool ScrubToStruct(const proto_processing_lib::proto_scrubber::ProtoScrubber* scrubber,
```
nit: Add a comment for this method?

dchakarwarti replied:
Done.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/scrubbing_util.cc:
```diff
@@ -413,7 +421,66 @@ IsMessageFieldPathPresent(const Protobuf::Type& type,
+
+  // Scrub the message.
+  absl::Status status = scrubber->Scrub(message);
+  if (!status.ok()) {
```
nit: log the status?

dchakarwarti replied:
Done.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/scrubbing_util.cc:
```diff
@@ -413,7 +421,66 @@ IsMessageFieldPathPresent(const Protobuf::Type& type,
+
+  // Convert the scrubbed message to proto.
+  status = ConvertToStruct(*message, type, type_helper.Resolver(), message_struct);
+  if (!status.ok()) {
```
nit: log the status?

dchakarwarti replied:
Done.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/scrubbing_util.cc:
```diff
@@ -413,7 +421,66 @@ IsMessageFieldPathPresent(const Protobuf::Type& type,
+  JsonParseOptions options;
+  auto status = Protobuf::util::JsonStringToMessage(cord_out_stream.Consume().Flatten(),
+                                                    message_struct, options);
+  if (!status.ok()) {
```
Log the original error message?

dchakarwarti replied:
Done.

shuoyang2016 commented on source/extensions/filters/http/proto_message_scrubbing/scrubbing_util/proto_scrubber.cc:
```diff
@@ -0,0 +1,205 @@
+  absl::StatusOr<std::string> status_or_target_resource =
+      ExtractStringFieldValue(*message_type_, type_finder_, field_mask.paths(0), message);
+  if (!status_or_target_resource.ok()) {
+    ENVOY_LOG_MISC(error, "Unable to extract target resource: {}",
```
Should this be error log?

dchakarwarti replied:
We can keep it as debug log.

dchakarwarti commented on source/extensions/filters/http/proto_message_scrubbing/filter.cc:
```diff
@@ -0,0 +1,434 @@
+                 "initiated when scrubbing proto messages");
+
+  auto buffering = request_msg_converter_->accumulateMessages(data, end_stream);
+  if (!buffering.ok()) {
```
Sure, created b/361256220 internally. We might need to add some util functionality common across these filters.

shuoyang2016 commented on CODEOWNERS:
```diff
@@ -35,8 +35,8 @@ extensions/filters/common/original_src @klarose @mattklein123
 /*/extensions/filters/http/grpc_field_extraction @taoxuy @nareddyt @yanavlasov
-# proto_message_logging http filter extension
-/*/extensions/filters/http/proto_message_logging @dchakarwarti @taoxuy @yanavlasov
+# proto_message_scrubbing http filter extension
```
I thought our plan is to use this for versioning/filtering scrubbing as well? @dchakarwarti  Sumit should have context.

dchakarwarti replied:
I confirmed with Sumit that the Proto Scrubber filter for visibility-related changes is separate from the PML/PMS filter and will be developed independently.

My rationale for renaming **Proto Message Logging** was that we aren't actually logging but rather writing extracted fields into dynamic metadata. However, I understand that the name **Proto Message Scrubbing** might be confusing.

Please let me know if you prefer to proceed with the `Proto Message Logging` filter as it is.

shuoyang2016 replied:
I see. Then Proto Message Extraction might be better? But would it confuse with gRPC field extraction? Proto Message Logging is indeed confusing too because it's not logging.

shuoyang2016 replied:
Didn't we think about merging this to gRPC field extraction filter?

TAOXUY replied:
> Didn't we think about merging this to gRPC field extraction filter?

Haven't we considered this already[1] and decide not to merge?

https://docs.google.com/document/d/1K7ANyvLkolM5o_pQP2hWpY0NJfI9rQ1czO1e5YaYzmo/edit?tab=t.0#heading=h.282wm8yq0rz6

dchakarwarti replied:
In that case, if we are looking for alternate names, how about:

1. Field Metadata Extractor
2. Proto Metadata Extractor
3. Proto Field Metadata Filter
4. Proto Field Capture

`Proto Metadata Extractor` might be the best balance between specificity and clarity.

I am okay if we decide to go with `Proto Message Extraction`.

Please let me know if any one sounds good.

shuoyang2016 replied:
I see. It's too long ago and I forgot about it. Then either way sounds good to me.

TAOXUY replied:
SG on Proto Message Extraction

dchakarwarti replied:
I have renamed the filter to `Proto Message Extraction` filter and thereby renamed necessary files and variables.

Thank you!

dchakarwarti commented on test/extensions/filters/http/proto_message_scrubbing/filter_test.cc:
```diff
@@ -0,0 +1,1837 @@
+  Envoy::Buffer::InstancePtr request_data = Envoy::Grpc::Common::serializeToGrpcFrame(request);
+
+  // Split into multiple buffers.
+  Envoy::Buffer::OwnedImpl start_request_data;
```
I think it is only used in `FilterTestExtractOk.UnaryMultipeBuffers` test case in `filter_test.cc` for extraction purpose. 

I have created a util method `splitBuffer()` that the request/response flow can use. Please let me know if I understand that correctly.

dchakarwarti commented on source/extensions/filters/http/proto_message_scrubbing/filter.cc:
```diff
@@ -0,0 +1,434 @@
+// Turns a '/package.service/method' to 'package.service.method' which is
+// the form suitable for the proto db lookup.
+absl::StatusOr<std::string> grpcPathToProtoPath(absl::string_view grpc_path) {
+  if (grpc_path.empty() || grpc_path.at(0) != '/' ||
```
I didn't find a similar functionality in the `common/grpc` code. However, the same function is used in [GFX filter](https://github.com/envoyproxy/envoy/blob/7344490c8240d341ba78c1f1205adac86a033a32/source/extensions/filters/http/grpc_field_extraction/filter.cc#L56).

We can reuse that as part of b/361256220 task.

dchakarwarti commented:
> It looks like tests do not provide sufficient coverage. Error: https://dev.azure.com/cncf/envoy/_build/results?buildId=178046&view=logs&j=4930ecaf-18f4-5b3c-dea3-309729c3b3ae&t=573d8780-d7b9-52e3-b4e0-a89886b0b9ff&l=5237
> 
> Cov report: https://storage.googleapis.com/envoy-pr/35689/coverage/index.html
> 
> /wait

I have added more unit tests. All presubmit tests seem to pass now.

Thank you!

dchakarwarti commented:
> Reviewed filter logic.
> 
> Will review scrubber logic shortly.

Thanks for the review, Shuo!

I have addressed the comments. Please let me know if it looks good.

Thank you!

yanavlasov commented:
I will wait for LGTMs from other reviewers and then submit the PR.

/wait

shuoyang2016 commented on docs/root/configuration/http/http_filters/proto_message_extraction_filter.rst:
```diff
@@ -0,0 +1,106 @@
+
+On the request and response path, it will check
+
+1. if the incoming gRPC request/response is configured, the filter tries to:
```
Could you document if the extraction is at critical path?

dchakarwarti replied:
Sure, I have updated the file regarding the same. Please let me know if anything needs clarification.

shuoyang2016 commented on source/extensions/filters/http/well_known_names.h:
```diff
@@ -50,6 +50,8 @@ class HttpFilterNameValues {
   // GRPC Field Extraction filter
   const std::string GrpcFieldExtraction = "envoy.filters.http.grpc_field_extraction";
+  // Proto Message Logging filter
+  const std::string ProtoMessageLogging = "envoy.filters.http.proto_scrubbing_logging";
```
Is this updated?

dchakarwarti replied:
Missed that. Updated it now.

Thanks for noticing!

shuoyang2016 reviewed (approved):
Thanks!

dchakarwarti commented:
/retest

dchakarwarti commented:
/retest-all

dchakarwarti commented:
> hi @dchakarwarti this is throwing a stack of warnings in our CI - probs the most important of them is the deprecation warnings on `cc_wkt_protos`
> 
> would it be possible to address these warnings upstream?
> 
> im also going to raise a pr with some minor docs cleanups

Hi Phlax!

Sure, this would require change in [Proto Processing Lib](https://github.com/grpc-ecosystem/proto_processing_lib). I will create a PR to update the SHA in envoy shortly.

Thanks!

dchakarwarti commented:
> > hi @dchakarwarti this is throwing a stack of warnings in our CI - probs the most important of them is the deprecation warnings on `cc_wkt_protos`
> > would it be possible to address these warnings upstream?
> > im also going to raise a pr with some minor docs cleanups
> 
> Hi Phlax!
> 
> Sure, this would require change in [Proto Processing Lib](https://github.com/grpc-ecosystem/proto_processing_lib). I will create a PR to update the SHA in envoy shortly.
> 
> Thanks!

JFYI: Created https://github.com/envoyproxy/envoy/pull/35921. Thanks!
